pkg/storage/s3.go

```diff
@@ -1,56 +1,318 @@
 	if options.Endpoint == "" && options.Region == "" {
 		return errors.New("must provide either 's3.region' or 's3.endpoint'")
 	}
-	// TODO: Verify Region.
+	if options.Endpoint != "" {
+		if !strings.HasPrefix(options.Endpoint, "https://") &&
+			!strings.HasPrefix(options.Endpoint, "http://") {
+			options.Endpoint = "http://" + options.Endpoint
+		}
+	}
+	if options.Provider == "alibaba" || options.Provider == "netease" ||
+		options.UseAccelerateEndpoint {
+		options.ForcePathStyle = false
+	}
+	if options.AccessKey == "" && options.SecretAccessKey != "" {
+		return errors.New("access_key not found")
+	}
+	if options.AccessKey != "" && options.SecretAccessKey == "" {
+		return errors.New("secret_access_key not found")
+	}
 
 	s3.Endpoint = options.Endpoint
 	s3.Region = options.Region
+	// StorageClass, SSE and ACL are acceptable to be empty
+	s3.StorageClass = options.StorageClass
+	s3.Sse = options.SSE
+	s3.Acl = options.ACL
+	s3.AccessKey = options.AccessKey
+	s3.SecretAccessKey = options.SecretAccessKey
+	s3.ForcePathStyle = options.ForcePathStyle
 	return nil
 }
 
 func defineS3Flags(flags *pflag.FlagSet) {
-	flags.String(s3EndpointOption, "", "Set the AWS S3 endpoint URL")
-	flags.String(s3RegionOption, "", "Set the AWS region")
-	// TODO: Finalize the list of options.
-	_ = flags.MarkHidden(s3EndpointOption)
+	flags.String(s3EndpointOption, "", "Set the S3 endpoint URL")
+	flags.String(s3RegionOption, "", "Set the S3 region")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs")
+	flags.String(s3ProviderOption, "", "Set the S3 provider")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	send, err := flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	if send {
+		options.AccessKey = os.Getenv("AWS_ACCESS_KEY_ID")
+		options.SecretAccessKey = os.Getenv("AWS_SECRET_ACCESS_KEY")
+	}
 	options.Endpoint, err = flags.GetString(s3EndpointOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	options.Region, err = flags.GetString(s3RegionOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
+	options.SSE, err = flags.GetString(s3SSEOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ACL, err = flags.GetString(s3ACLOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.StorageClass, err = flags.GetString(s3StorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ForcePathStyle = true
+	options.Provider, err = flags.GetString(s3ProviderOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	return options, err
+}
+
+// newS3Storage initialize a new s3 storage for metadata
+func newS3Storage(s3Back *backup.S3) (*S3Storage, error) {
```

Comment:
The name `s3Back` is weird. What about using `backend`?

Reply:
done